dtm0/fop.c

```diff
@@ -341,50 +349,92 @@ M0_INTERNAL void m0_dtm0_on_committed(struct m0_fom                *fom,
-		       !!m0_dtm0_is_a_volatile_dtm(fom->fo_service),
-		       FID_P(&fom->fo_service->rs_service_fid),
-		       fom->fo_service->rs_reqh);
-
-		if (m0_dtm0_in_ut() && req->dtr_msg == DMT_EXECUTE &&
-		    m0_dtm0_is_a_persistent_dtm(fom->fo_service)) {
-			rc = m0_dtm0_tx_desc_copy(&req->dtr_txr,
-						  &rep->dr_txr);
-			M0_ASSERT(rc == 0);
-			m0_dtm0_send_msg(fom, DMT_EXECUTED,
-					 &req->dtr_txr.dtd_id.dti_fid,
-					 &req->dtr_txr);
-		}
+		break;
 
+	case M0_FOPH_DTM0_PREP:
 		if (req->dtr_msg == DTM_PERSISTENT &&
-		    m0_dtm0_is_a_volatile_dtm(fom->fo_service)) {
-			/* TODO: Only the client side is here so far.
-			 * Remove the "is_volatile" part when plog is ready.
-			 */
-			m0_be_dtm0_log_pmsg_post(svc->dos_log, fom->fo_fop);
+			m0_dtm0_is_a_persistent_dtm(fom->fo_service)) {
+			rc = m0_dtm0_tx_desc_copy(&req->dtr_txr, &txd);
+			if (rc != 0) {
+				m0_fom_phase_move(fom, M0_ERR(rc), M0_FOPH_FAILURE);
+			}
+
+			m0_be_dtm0_log_credit(M0_DTML_PERSISTENT,
+								  &txd,
+								  &buf,
+								  m0_fom_reqh(fom)->rh_beseg,
+								  NULL,
+								  &dtm0logrec_cred);
+			m0_be_tx_credit_add(&fom->fo_tx.tx_betx_cred, &dtm0logrec_cred);
+			rc = m0_be_tx_open_sync(&fom->fo_tx.tx_betx);
+			if (rc != 0) {
+				m0_fom_phase_move(fom, M0_ERR(rc), M0_FOPH_FAILURE);
+			}
 		}
-
-		rep->dr_rc = 0;
-		m0_fom_phase_set(fom, M0_FOPH_SUCCESS);
-		rc = M0_FSO_AGAIN;
+		m0_fom_phase_set(fom, M0_FOPH_DTM0_LOGGING);
+		break;
+
+case M0_FOPH_DTM0_LOGGING:
+rep = m0_fop_data(fom->fo_rep_fop);
+M0_SET0(&rep->dr_txr);
+svc = m0_dtm0_service_find(fom->fo_service->rs_reqh);
+M0_ASSERT(svc != NULL);
+
+M0_LOG(M0_DEBUG, "Processing non-generic phase %" PRIu32
+" %d, " FID_F ", %p", req->dtr_msg,
+!!m0_dtm0_is_a_volatile_dtm(fom->fo_service),
+FID_P(&fom->fo_service->rs_service_fid),
+			fom->fo_service->rs_reqh);
+
+			if (req->dtr_msg == DMT_EXECUTE &&
+					m0_dtm0_is_a_persistent_dtm(fom->fo_service)) {
+				M0_ASSERT(m0_dtm0_in_ut());
+				rc = m0_dtm0_tx_desc_copy(&req->dtr_txr, &rep->dr_txr);
+				if (rc != 0) {
+					m0_fom_phase_move(fom, M0_ERR(rc), M0_FOPH_FAILURE);
+				}
+				m0_dtm0_send_msg(fom, DMT_EXECUTED,
+								 &req->dtr_txr.dtd_id.dti_fid,
+								 &req->dtr_txr);
+			} else if (req->dtr_msg == DTM_PERSISTENT) {
+				if (m0_dtm0_is_a_volatile_dtm(fom->fo_service)) {
+								m0_be_dtm0_log_pmsg_post(svc->dos_log, fom->fo_fop);
+				} else {
+					m0_mutex_lock(&svc->dos_log->dl_lock);
+					m0_dtm0_logrec_update(svc->dos_log, &fom->fo_tx.tx_betx, &txd, &buf);
+					m0_mutex_unlock(&svc->dos_log->dl_lock);
+					m0_be_tx_close_sync(&fom->fo_tx.tx_betx);
+				}
+			}
+
+			rep->dr_rc = 0;
+			m0_fom_phase_set(fom, M0_FOPH_SUCCESS);
+			rc = M0_FSO_AGAIN;
+			break;
+
+		default:
+			M0_IMPOSSIBLE("Invalid phase");
 	}
 
+	phase_sm_id = m0_sm_id_get(&fom->fo_sm_phase);
+	rpc_sm_id   = m0_sm_id_get(&item->ri_sm);
+	M0_ADDB2_ADD(M0_AVI_FOM_TO_TX, phase_sm_id, rpc_sm_id);
 	return M0_RC(rc);
```

Comment:
does it always return M0_FSO_WAIT/AGAIN?
For example, after line 391 there's no rc=AGAIN assignment.

Reply:
Thanks. Fixed.